lib/vector-tap/src/lib.rs

```diff
@@ -171,57 +183,194 @@ impl<'a> TapRunner<'a> {
+                {
+                    return Ok(())
                 }
                 Ok(_) => return Err(TapExecutorError::GraphQLError),
             }
         }
     }
+
+    #[allow(clippy::print_stdout)]
+    fn output_event_stdout(
+        &self,
+        output_events: &[GraphQLTapOutputEvent],
+        formatter: &EventFormatter,
+    ) {
+        for tap_event in output_events.iter() {
+            #[allow(clippy::print_stdout)]
+            #[allow(clippy::print_stderr)]
+            match tap_event {
+                GraphQLTapOutputEvent::Log(ev) => {
+                    println!(
+                        "{}",
+                        formatter.format(
+                            ev.component_id.as_ref(),
+                            ev.component_kind.as_ref(),
+                            ev.component_type.as_ref(),
+                            ev.string.as_ref()
+                        )
+                    );
+                }
+                GraphQLTapOutputEvent::Metric(ev) => {
+                    println!(
+                        "{}",
+                        formatter.format(
+                            ev.component_id.as_ref(),
+                            ev.component_kind.as_ref(),
+                            ev.component_type.as_ref(),
+                            ev.string.as_ref()
+                        )
+                    );
+                }
+                GraphQLTapOutputEvent::Trace(ev) => {
+                    println!(
+                        "{}",
+                        formatter.format(
+                            ev.component_id.as_ref(),
+                            ev.component_kind.as_ref(),
+                            ev.component_type.as_ref(),
+                            ev.string.as_ref()
+                        )
+                    );
+                }
+                GraphQLTapOutputEvent::EventNotification(ev) => {
+                    eprintln!("{}", ev.message);
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use chrono::Utc;
+    use futures_util::sink::SinkExt;
+    use futures_util::stream::StreamExt;
+    use serde_json::Value;
+    use tokio::net::TcpListener;
+    use tokio::time::sleep;
+    use tokio_tungstenite::accept_async;
+    use tokio_tungstenite::tungstenite::Message;
+
+    use portpicker::pick_unused_port;
+
+    use super::*;
+
+    #[tokio::test(start_paused = true)]
+    async fn test_async_output_channel() {
+        let component_id = "test-component-id";
+        let component_type = "test-component-type";
+        let message = "test-message";
+        let timestamp = Utc::now();
+        let string_encoding = "test-str";
+
+        // Start a local WebSocket server to mimic Vector GraphQL API
+        let ip_addr = IpAddr::V4(Ipv4Addr::LOCALHOST);
+        let port = pick_unused_port(ip_addr);
+        let addr = format!("{}:{}", ip_addr, port);
```

Comment:
Nit: 
```suggestion
        let addr = format!("{ip_addr}:{port}");
```

Basically you can inline the variable in `format!`.